sql/group.go

```diff
@@ -0,0 +1,438 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+var aggregates = map[string]aggregateFunc{
+	"avg":   &avgAggregate{},
+	"count": &countAggregate{},
+	"max":   &maxAggregate{},
+	"min":   &minAggregate{},
+	"sum":   &sumAggregate{},
+}
+
+func (p *planner) groupBy(n *parser.Select, s *scanNode) (*groupNode, error) {
+	// TODO(pmattis): This only handles aggregate functions, not GROUP BY.
+
+	// Loop over the render expressions and extract any aggregate functions.
+	var funcs []*parser.FuncExpr
+	for _, r := range s.render {
+		f, err := extractAggregateFuncs(r)
+		if err != nil {
+			return nil, err
+		}
+		funcs = append(funcs, f...)
+	}
+	if len(funcs) == 0 {
+		return nil, nil
+	}
+
+	// Aggregation is being performed. Loop over the render expressions again and
+	// verify that the only qvalues mentioned by GROUP BY expressions are allowed
+	// outside of the aggregate function arguments. For example, the following is
+	// illegal because k is used outside of the aggregate function and not part
+	// of a GROUP BY expression.
+	//
+	//   SELECT COUNT(k), k FROM kv GROUP BY v
```

Comment:
And in sqlite as well. The sqlite documentation says it selects an arbitrary row for the value of `k`.